### DISCUSSION TITLE
TypeScript Migration

### DISCUSSION THREAD

**Discussion by Zemledelec** [openglobus/openglobus#757] [Announcements]
Hey, guys @openglobus/contributors, @program-sam! I've started typescript migration. Migration is going smoothly. Somewhere I used `any` declaration, a few times I used `@ts-ignore`, and a few times type assertion, but nothing critical in the code design. Check it out and give feedback!

PR https://github.com/openglobus/openglobus/pull/756

**Comment by tbo47** [comment]
I see you removed the examples. Are they going to be somewhere else? Do you need help to bring they back. The examples could be in javascript, they don't need to be in typescript. Do you need help with bringing the examples back?

**Reply by Zemledelec** [reply]
Hi, @tbo47! Yes, right, I removed examples and created a new branch in the examples repository over here.

There are some thoughts:

Examples branch: https://github.com/openglobus/examples/tree/typescript_migration

First of all, I believe, that it would be possible to use openglobus both as a javascript or typescript library. I see three ways:
1) As **esm** module (in the typescript migration branch sandbox/osm https://github.com/openglobus/openglobus/tree/657_typescript_migration), I use `npm run dev` for rollup watcher and builder.
2) As npm package(I suppose it's the same as esm but with npm), it's already here, but no typescripted yet.
3) Also, there is a possibility to use the library as a javascript modules from `./out-tsc` folder which we can make with `tsc` command.

I think that the most convenient way which leads to understand "how it works" (example feature), and how to use the library as a third-party tool - is to use the library as `esm` module. We can organize the module as we want to, all functionality is available inside index.ts in the src root folder.

So, I propose to create a submodule with the openglobus library with esm, as I described in (1) (npm is not available for typescript right now), and try to launch examples with the new way, with the esm module.

It would be a great help if you prepare the examples! Some modules are not available with the new way, such as `control`, might be something else...and there are a lot of js still remaining. So if you try to compile it as a typescript you will get a lot of ts errors. But if you use it as a javascript everything will be right. 

Could you please check the code and tell what you think about how to make examples js, ts, or both maybe?

**Reply by pavletto** [reply]
But we thought to get rid of the examples, right?
I already start branch with examples in sandbox-style in renew_examples in same repo

**Reply by tbo47** [reply]
@pavletto I let you finish what you already started.

**Reply by Zemledelec** [reply]
@pavletto I already removed examples to a separate repository over here https://github.com/openglobus/examples/tree/typescript_migration. Because typescript library is not compatible with the current examples.

**Reply by Zemledelec** [reply]
There must be only one type of examples, which uses esm module. It could be a submodule, or npm package, it doesn't matter. But the main idea it must be only one separate example repo. And old (current) version of examples, which I replaced to the repo is the easiest to refactor, because all imports already exists.

**Reply by Zemledelec** [reply]
I want to simplify it as much as possible. For example, no need for external components or a framework to maintain examples. Just clone and run each example in a browser, like `npm run serve`, and voila, `127.0.0.1:8080/attribution/attribution.html` like that. This is a base state, and It should be compatible with your solution. If it's so, I agree. More to say it will be very useful when it will be embedded in the homepage site.

**Reply by Zemledelec** [reply]
For example, as I see, there are two repos: examples - with bare examples as it is, and sandbox - a framework, page site which serves or provides examples as a list of examples, with one side scree - code, and other - beautiful result.

**Reply by Zemledelec** [reply]
Here it is, https://github.com/openglobus/openglobus/issues/685, it's better to add the number of an issue in a branch name like `685_renew_examples`, and create a new issue in the example repository, first.

**Comment by tbo47** [comment]
You should remove `npm test` or `npm run test` in the `./.github` directory for now. I can add the tests back in a later PR.

**Reply by Zemledelec** [reply]
I launch tests time to time (it's just need to remove .js), I could do it. No need to worry about I'll fix it.

**Comment by pavletto** [comment]
Proposale: maybe we can remove additional nesting level in src directory? I mean `src/og` folder. and move all files to src directly?

**Reply by Zemledelec** [reply]
Agree, I'll do it after migration. Right now it is not critical.

**Comment by wewindy** [comment]
Congratulations

**Comment by Zemledelec** [comment]
DONE!
